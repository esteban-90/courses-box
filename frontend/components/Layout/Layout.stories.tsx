import { ComponentStoryObj, ComponentMeta } from '@storybook/react'
import { expect } from '@storybook/jest'
import { screen } from '@storybook/testing-library'
import { Layout } from '@/components/Layout'

export default {
  title: 'Content/Layout',
  component: Layout,
} as ComponentMeta<typeof Layout>

const child = (
  <>
    <h1>Main article area</h1>
    <p>
      In this layout, we display the areas in source order for any screen less than 500 pixels wide. We go to a two
      column layout, and then to a three column layout by redefining the grid,and the placement of items on the grid.
    </p>
  </>
)

export const BasicLayout: ComponentStoryObj<typeof Layout> = {
  async play() {
    await expect(screen.getByText('Main article area')).toBeInTheDocument()
  },

  args: {
    children: child,
  },
}
