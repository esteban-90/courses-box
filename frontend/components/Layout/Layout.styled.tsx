import styled from '@emotion/styled'
import { Input as _Input } from '@/components/Input'
import { Link } from '@/components/Link'
import { Logo as _Logo } from '@/components/Logo'

export const Content = styled.main`
  grid-area: content;
  min-height: 84vh;
  margin-top: 1rem;
`

export const Footer = styled.footer`
  grid-area: footer;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
`

export const Input = styled(_Input)`
  grid-area: search;
  width: 100%;
  height: 3.5rem;
`

export const Logo = styled(_Logo)`
  grid-area: header;
  display: flex;
  justify-content: flex-start;
  align-items: center;
  height: 3.5rem;

  & .logo_full {
    display: none;
  }

  @media (min-width: 560px) {
    & .logo_short {
      display: none;
    }

    & .logo_full {
      display: inline;
    }
  }
`

export const LogoLink = styled(Link)`
  padding-right: 1vw;
`

export const Navigation = styled.nav`
  grid-area: nav;
  display: flex;
  justify-content: space-around;
  align-items: center;
  margin: 0 2vmin;
`

export const Wrapper = styled.div`
  display: grid;
  gap: 0.1rem;
  padding: 0.5rem;
  color: ${({ theme }) => theme.font.regular};
  background-color: ${({ theme }) => theme.background};

  grid-template-areas:
    'header nav'
    'search search'
    'content content'
    'footer footer';

  nav {
    flex-direction: row;
    justify-content: flex-end;
    gap: 5vmin;
  }

  @media (min-width: 500px) {
    grid-template-columns: 1fr 3fr;
  }

  @media (min-width: 960px) {
    grid-template-columns: 1fr 4fr 2fr;

    grid-template-areas:
      'header search nav'
      'content content content'
      'footer footer footer';
  }
`
