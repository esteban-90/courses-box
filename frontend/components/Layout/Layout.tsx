import { FC, useState, useEffect, useLayoutEffect } from 'react'
import { ThemeProvider } from '@emotion/react'
import NextLink from 'next/link'
import { IconButton } from '@/components/IconButton'
import { Link } from '@/components/Link'
import { Themes } from '@/styles/themes'
import { LayoutProps } from '@/types'
import * as Styled from './Layout.styled'

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect

export const Layout: FC<LayoutProps> = ({ children }): JSX.Element => {
  const [isDark, setIsDark] = useState(true)
  const theme = Themes[isDark ? 'dark' : 'light']

  const toggleTheme = () => {
    localStorage.setItem('courses-box-theme', isDark ? 'light' : 'dark')
    setIsDark(!isDark)
  }

  useIsomorphicLayoutEffect(() => {
    const savedTheme = localStorage.getItem('courses-box-theme')
    const savedThemeExists = savedTheme !== null
    const savedThemeIsDark = savedTheme === 'dark'
    const preferenceIsDark = window.matchMedia('prefers-color-scheme: dark').matches
    const themeIsDark = savedThemeExists ? savedThemeIsDark : preferenceIsDark

    setIsDark(themeIsDark)
  }, [])

  return (
    <ThemeProvider theme={theme}>
      <Styled.Wrapper>
        <NextLink href='/' passHref>
          <Styled.LogoLink>
            <Styled.Logo size={3}>
              <span className='logo_short'>C8X</span>
              <span className='logo_full'>CoursesBox</span>
            </Styled.Logo>
          </Styled.LogoLink>
        </NextLink>

        <Styled.Navigation>
          <NextLink href='/all' passHref>
            <Link>All</Link>
          </NextLink>

          <NextLink href='/login' passHref>
            <IconButton name='Login' size={1} />
          </NextLink>

          <IconButton name={isDark ? 'Sun' : 'Moon'} size={1} onClick={toggleTheme} />
        </Styled.Navigation>

        <Styled.Input icon='Search' placeholder='Search' onChange={() => {}} />
        <Styled.Content>{children}</Styled.Content>

        <Styled.Footer>
          <p>
            &copy; {new Date().getFullYear()}&nbsp;
            <Link href='https://github.com/esteban-90' target='_blank'>
              Esteban V.M.
            </Link>
            &nbsp;
          </p>
          <p>All rights reserved</p>
        </Styled.Footer>
      </Styled.Wrapper>
    </ThemeProvider>
  )
}
