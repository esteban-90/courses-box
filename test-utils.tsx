import { FC, ReactElement, ReactNode } from 'react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Themes } from '@/styles/themes'

const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
  return <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions): RenderResult => {
  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render }

// 14
