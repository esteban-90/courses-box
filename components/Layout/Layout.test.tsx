import { render } from '@/test-utils'
import { Layout } from './Layout'

describe('Layout test cases', () => {
  const child = (
    <>
      <h1>Main article area</h1>
      <p>
        In this layout, we display the areas in source order for any screen less than 500 pixels wide. We go to a two
        column layout, and then to a three column layout by redefining the grid,and the placement of items on the grid.
      </p>
    </>
  )

  it('Render check', () => {
    const { asFragment } = render(
      <Layout
        onThemeToggle={function (): never {
          throw new Error('Function not implemented.')
        }}
        isDark
      >
        {child}
      </Layout>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
