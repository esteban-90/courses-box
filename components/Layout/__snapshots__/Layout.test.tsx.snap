// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout test cases Render check 1`] = `
<DocumentFragment>
  <div
    class="css-1qtm68j"
  >
    <a
      class="css-dj4799"
      href="/"
    >
      <header
        class="css-1l7s6s"
        size="3"
      >
        <span
          class="logo_short"
        >
          C8X
        </span>
        <span
          class="logo_full"
        >
          CoursesBox
        </span>
      </header>
    </a>
    <nav
      class="css-mpfgc5"
    >
      <a
        class="css-1v1fc5j"
        href="/all"
      >
        All
      </a>
      <a
        class="css-1v1fc5j"
        href="/news"
      >
        News
      </a>
      <button
        class="css-1aqtdpz"
        size="2"
        title="Sun"
      >
        <svg
          aria-label="Sun"
          class="css-z3g3v6"
          fill="currentColor"
          height="1rem"
          role="img"
          viewBox="0 0 16 16"
          width="1rem"
        >
          <path
            d="M8 11a3 3 0 110-6 3 3 0 010 6zm0 1a4 4 0 100-8 4 4 0 000 8zM8 0a.5.5 0 01.5.5v2a.5.5 0 01-1 0v-2A.5.5 0 018 0zm0 13a.5.5 0 01.5.5v2a.5.5 0 01-1 0v-2A.5.5 0 018 13zm8-5a.5.5 0 01-.5.5h-2a.5.5 0 010-1h2a.5.5 0 01.5.5zM3 8a.5.5 0 01-.5.5h-2a.5.5 0 010-1h2A.5.5 0 013 8zm10.657-5.657a.5.5 0 010 .707l-1.414 1.415a.5.5 0 11-.707-.708l1.414-1.414a.5.5 0 01.707 0zm-9.193 9.193a.5.5 0 010 .707L3.05 13.657a.5.5 0 01-.707-.707l1.414-1.414a.5.5 0 01.707 0zm9.193 2.121a.5.5 0 01-.707 0l-1.414-1.414a.5.5 0 01.707-.707l1.414 1.414a.5.5 0 010 .707zM4.464 4.465a.5.5 0 01-.707 0L2.343 3.05a.5.5 0 11.707-.707l1.414 1.414a.5.5 0 010 .708z"
          />
        </svg>
      </button>
    </nav>
    <label
      class="css-1fh6xqn"
      height="7"
      width="20"
    >
      <div
        class="css-6nw0cx"
      >
        <input
          class="css-1cvyunr"
          placeholder="Search"
        />
        <svg
          aria-label="Search"
          class="css-5n1pld"
          fill="currentColor"
          height="2rem"
          role="img"
          viewBox="0 0 16 16"
          width="2rem"
        >
          <path
            d="M11.742 10.344a6.5 6.5 0 10-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 001.415-1.414l-3.85-3.85a1.007 1.007 0 00-.115-.1zM12 6.5a5.5 5.5 0 11-11 0 5.5 5.5 0 0111 0z"
          />
        </svg>
      </div>
    </label>
    <main
      class="css-1qc1fvq"
    >
      <h1>
        Main article area
      </h1>
      <p>
        In this layout, we display the areas in source order for any screen less than 500 pixels wide. We go to a two column layout, and then to a three column layout by redefining the grid,and the placement of items on the grid.
      </p>
    </main>
    <footer
      class="css-1y0dtms"
    >
      Â© 2022 Esteban V.M. All rights reserved
    </footer>
  </div>
</DocumentFragment>
`;
